Fix autograd engine checks and update InputMetadata

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65235

1. Updated the legacy type checks in `torch/csrc/autograd/engine.cpp` to individually validate the dtype, device, and layout equality for grad and tensor.
2. Removed device field from `InputMetadata` since it's already stored via storing options. Also, added `dtype()` and `layout()` methods to `InputMetadata`. To make this change, some calls had to be updated due to the change in constructor.
3. To fix https://github.com/pytorch/pytorch/issues/65016:
     a. Added a `is_tensor_subclass` field in `InputMetadata` to skip device checks for grad and tensor when the tensor has 
         python key set on it (tensor subclass).

Differential Revision: [D31117318](https://our.internmc.facebook.com/intern/diff/D31117318)

anjali411 committed:
Fix autograd engine checks and update InputMetadata

[ghstack-poisoned]

anjali411 committed:
Update on "Fix autograd engine checks and update InputMetadata"




[ghstack-poisoned]

anjali411 committed:
Update on "Fix autograd engine checks and update InputMetadata"




[ghstack-poisoned]

albanD commented on torch/csrc/autograd/input_metadata.h:
```diff
@@ -11,22 +11,29 @@
 /**
- * Records type, shape, and device of tensor and, where applicable,
- * the stream the correspondingoperation took place on.
+ * Records TensorOptions, and shape of the tensor and, where applicable,
```
also is python dispatch no?

albanD commented on torch/csrc/autograd/input_metadata.h:
```diff
@@ -11,22 +11,29 @@
+    stream_ = c10::impl::getDeviceGuardImpl(device_.type())->getStream(device_);
+  }
+
+  InputMetadata(const at::TensorOptions options, at::IntArrayRef shape)
```
When is this constructor called?
How is the python dispatch initialized?

anjali411 replied:
removed it

albanD commented on test/test_python_dispatch.py:
```diff
@@ -449,5 +449,12 @@ def test_nested_enable_python_mode(self) -> None:
+        x = LoggingTensor(torch.tensor([3.0, 4.0], requires_grad=True))
+        y = torch.randn(2)
+        z = x * y
+        assert isinstance(z, LoggingTensor)
```
self.assertIsInstance

albanD commented on test/test_python_dispatch.py:
```diff
@@ -449,5 +449,12 @@ def test_nested_enable_python_mode(self) -> None:
+        y = torch.randn(2)
+        z = x * y
+        assert isinstance(z, LoggingTensor)
+        z.sum().backward(torch.tensor(1.))
```
Can you check that the computed gradient is correct?

ezyang commented:
I'm letting Alban do full review on this.

anjali411 commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -701,22 +695,31 @@ void validate_outputs(
        std::stringstream ss;
-       ss << "invalid gradient at index " << i << " - expected type ";
-       ss << metadata.options() << " but got " << grad.options();
+       ss << "invalid gradient at index " << i << " - expected layout ";
```
There are currently four layouts: sparse, sparsecsr, mkldnn and strided. Currently, the only cross combination of layouts for tensor and grad that's supported is `(Strided, Sparse)` (based on the check that was being done by `is_compatible_type`). I am not sure about mkldn but it seems like we should support all possible combinations of (tensor.layout(), tensor.grad.layout()) for the other three layouts. cc. @albanD

albanD replied:
I think we can keep this current layout logic as-is and only update the subclass case.

anjali411 committed:
Update on "Fix autograd engine checks and update InputMetadata"


1. Updated the legacy type checks in `torch/csrc/autograd/engine.cpp` to individually validate the dtype, device, and layout equality for grad and tensor.
2. Removed device field from `InputMetadata` since it's already stored via storing options. Also, added `dtype()` and `layout()` methods to `InputMetadata`. To make this change, some calls had to be updated due to the change in constructor.
3. To fix https://github.com/pytorch/pytorch/issues/65016:
     a. Added a `is_tensor_subclass` field in `InputMetadata` to skip device checks for grad and tensor when the tensor has 
         python key set on it (tensor subclass).

[ghstack-poisoned]

anjali411 committed:
Update on "Fix autograd engine checks and update InputMetadata"


1. Updated the legacy type checks in `torch/csrc/autograd/engine.cpp` to individually validate the dtype, device, and layout equality for grad and tensor.
2. Removed device field from `InputMetadata` since it's already stored via storing options. Also, added `dtype()` and `layout()` methods to `InputMetadata`. To make this change, some calls had to be updated due to the change in constructor.
3. To fix https://github.com/pytorch/pytorch/issues/65016:
     a. Added a `is_tensor_subclass` field in `InputMetadata` to skip device checks for grad and tensor when the tensor has 
         python key set on it (tensor subclass).

[ghstack-poisoned]

albanD reviewed (commented):
sgtm just the continue thing

anjali411 committed:
Update on "Fix autograd engine checks and update InputMetadata"


1. Updated the legacy type checks in `torch/csrc/autograd/engine.cpp` to individually validate the dtype, device, and layout equality for grad and tensor.
2. Removed device field from `InputMetadata` since it's already stored via storing options. Also, added `dtype()` and `layout()` methods to `InputMetadata`. To make this change, some calls had to be updated due to the change in constructor.
3. To fix https://github.com/pytorch/pytorch/issues/65016:
     a. Added a `is_tensor_subclass` field in `InputMetadata` to skip device checks for grad and tensor when the tensor has 
         python key set on it (tensor subclass).

[ghstack-poisoned]

albanD reviewed (approved):
Sounds good!

anjali411 committed:
Update on "Fix autograd engine checks and update InputMetadata"


1. Updated the legacy type checks in `torch/csrc/autograd/engine.cpp` to individually validate the dtype, device, and layout equality for grad and tensor.
2. Removed device field from `InputMetadata` since it's already stored via storing options. Also, added `dtype()` and `layout()` methods to `InputMetadata`. To make this change, some calls had to be updated due to the change in constructor.
3. To fix https://github.com/pytorch/pytorch/issues/65016:
     a. Added a `is_tensor_subclass` field in `InputMetadata` to skip device checks for grad and tensor when the tensor has 
         python key set on it (tensor subclass).

Differential Revision: [D31082693](https://our.internmc.facebook.com/intern/diff/D31082693)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31082693).

anjali411 committed:
Update on "Fix autograd engine checks and update InputMetadata"


1. Updated the legacy type checks in `torch/csrc/autograd/engine.cpp` to individually validate the dtype, device, and layout equality for grad and tensor.
2. Removed device field from `InputMetadata` since it's already stored via storing options. Also, added `dtype()` and `layout()` methods to `InputMetadata`. To make this change, some calls had to be updated due to the change in constructor.
3. To fix https://github.com/pytorch/pytorch/issues/65016:
     a. Added a `is_tensor_subclass` field in `InputMetadata` to skip device checks for grad and tensor when the tensor has 
         python key set on it (tensor subclass).

Differential Revision: [D31082693](https://our.internmc.facebook.com/intern/diff/D31082693)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31117318).

anjali411 commented:
@ezyang @albanD should we cherry-pick this in 1.10?

albanD commented:
I would say no since I don't think we plan to have broad subclassing support in 1.10 anyways?
